File: 1256-rank-transform-of-an-array/rank-transform-of-an-array.java
class Solution {
    public int[] arrayRankTransform(int[] arr) {
        int[] sortedArray = arr.clone();
        Arrays.sort(sortedArray);

        int uniqueCount = 0;
        for(int i = 0; i < arr.length; i++){
            if(i == 0 || sortedArray[i - 1] != sortedArray[i]){
                sortedArray[uniqueCount] = sortedArray[i];
                uniqueCount++;
            }
        }

        int[] result = new int[arr.length];
        for(int i = 0; i < arr.length; i++){
            result[i] = Arrays.binarySearch(sortedArray, 0, uniqueCount, arr[i]) + 1;
        }
        return result;
    }
}

File: 1256-rank-transform-of-an-array/rank-transform-of-an-array.py
class Solution:
    def arrayRankTransform(self, arr: List[int]) -> List[int]:
        sortedArr = sorted(set(arr))
        result = []
        for num in arr:
            result.append(self.binarySearch(0, len(sortedArr), sortedArr, num) + 1)

        return result

        
    
    def binarySearch(self, l, r, arr, target):
        while(l <= r):
            mid = l + (r - l)//2
            if arr[mid] == target:
                return mid
            elif arr[mid] < target:
                l = mid + 1
            else: r = mid - 1
        